### ISSUE TITLE
[1.12, 1.13] Regression in enum narrowing

### ISSUE THREAD

**Issue by Fred-si** [python/mypy#18029]
**Bug Report**

Regression in 1.12.

Try to narrowing union between StrEnum (or IntEnum) and Literal no longer works since mypy 1.12.0

**To Reproduce**
Issue can be reproduce with this code sample
```python
from typing_extensions import Literal
from typing import Union
from enum import StrEnum

class Foo(StrEnum):
    FOO = 'a'


def foo(a: Union[Foo, Literal['foo']]) -> Foo:
    if a == 'foo':
        return Foo.FOO

    return a
```
[playground link](https://mypy-play.net/?mypy=1.12.0&python=3.12&gist=19f4a8eaf079d73e704b4bb42e48b2fd)
failing test can be found [here](https://github.com/Fred-si/mypy/blob/d3b4ef5e5abbf74f7a7002423e50175743f898bf/test-data/unit/check-literal.test#L2988-L3017)

**Expected Behavior**
After the if branch return, variable type should be narrowed to Foo and mypy should not produce error

**Actual Behavior**
After the if branch return, variable type is considered as "Union[Foo, Literal['foo']]" and my

**Your Environment**
- Mypy version used: 1.12, 1.13, master
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: all versions from 3.8 to 3.13

**Comment by JukkaL** [comment]
This bisects to #17866. I'll look into this.

**Comment by JukkaL** [comment]
Narrowing has the issue already in 1.11, but the recent changes made the issue more prominent. This doesn't work as expected when using mypy 1.11:
```py
from typing_extensions import Literal
from typing import Union

class C:
    def __eq__(self, other): return True

def f(x: Union[C, Literal['foo']]) -> None:
    if x == "foo":
        reveal_type(x)  # Union[__main__.C, Literal['foo']]
    else:
        reveal_type(x)  # Union[__main__.C, Literal['foo']]
```
